network/discovery/discovery.go

```diff
@@ -0,0 +1,427 @@
+// and connects them to the current running node
+type DiscoveryService struct {
+	proto.UnimplementedDiscoveryServer
+
+	baseServer   networkingServer // The interface towards the base networking server
+	logger       hclog.Logger     // The DiscoveryService logger
+	routingTable *kb.RoutingTable // Kademlia 'k-bucket' routing table that contains connected nodes info
+
+	closeCh chan struct{} // Channel used for stopping the DiscoveryService
+}
+
+// NewDiscoveryService creates a new instance of the discovery service
+func NewDiscoveryService(
+	server networkingServer,
+	routingTable *kb.RoutingTable,
+	logger hclog.Logger,
+	closeCh chan struct{},
+) *DiscoveryService {
+	return &DiscoveryService{
+		logger:       logger.Named("discovery"),
+		baseServer:   server,
+		routingTable: routingTable,
+		closeCh:      closeCh,
+	}
+}
+
+// Start starts the discovery service
+func (d *DiscoveryService) Start() {
+	go d.startDiscovery()
+}
+
+// Close stops the discovery service
+func (d *DiscoveryService) Close() {
+	close(d.closeCh)
+}
+
+// RoutingTableSize returns the size of the routing table
+func (d *DiscoveryService) RoutingTableSize() int {
+	if d.routingTable == nil {
+		return 0
+	}
+
+	return d.routingTable.Size()
+}
+
+// RoutingTablePeers fetches the peers from the routing table
+func (d *DiscoveryService) RoutingTablePeers() []peer.ID {
+	if d.routingTable == nil {
+		return []peer.ID{}
+	}
+
+	return d.routingTable.ListPeers()
+}
+
+// HandleNetworkEvent handles base network events for the DiscoveryService
+func (d *DiscoveryService) HandleNetworkEvent(peerEvent *event.PeerEvent) {
+	peerID := peerEvent.PeerID
+
+	switch peerEvent.Type {
+	case event.PeerConnected:
+		// Add peer to the routing table and to our local peer table
+		_, err := d.routingTable.TryAddPeer(peerID, false, false)
+		if err != nil {
+			d.logger.Error("failed to add peer to routing table", "err", err)
+
+			return
+		}
+	case event.PeerDisconnected, event.PeerFailedToConnect:
+		// Run cleanup for the local routing / reference peers table
+		d.routingTable.RemovePeer(peerID)
+	}
+}
+
+// ConnectToBootnodes attempts to connect to the bootnodes
+// and add them to the peer / routing table
+func (d *DiscoveryService) ConnectToBootnodes(bootnodes []*peer.AddrInfo) {
+	for _, nodeInfo := range bootnodes {
+		if err := d.addToTable(nodeInfo); err != nil {
+			d.logger.Error(
+				"Failed to add new peer to routing table",
+				"peer",
+				nodeInfo.ID,
+				"err",
+				err,
+			)
+		}
+	}
+}
+
+// addToTable adds the node to the peer store and the routing table
+func (d *DiscoveryService) addToTable(node *peer.AddrInfo) error {
+	// before we include peers on the routing table -> dial queue
+	// we have to add them to the peer store so that they are
+	// available to all the libp2p services
+	d.baseServer.AddToPeerStore(node)
+
+	if _, err := d.routingTable.TryAddPeer(
+		node.ID,
+		false,
+		false,
+	); err != nil {
+		return err
+	}
```

Comment:
As we discussed, it might be best to reverse the order of these 2 calls (if that wouldn't break anything). In case of a failing `TryAddPeer`, the peer will still be added to the peerstore, yet it will not be preset in the routing table.

Reply:
Agreed 👍 

I've added rollback support, as well as a unit test covering both scenarios (since in fact this peer store addition needs to be present before adding it to the routing table):

[25e424f](https://github.com/0xPolygon/polygon-edge/pull/343/commits/25e424fc14a3708a1704165796bbe8f41392da74)